### PR TITLE
Headers

### PR DESCRIPTION
**Pull request by antmat**
http2-like headers implementation

### CODE REVIEW

File: include/cocaine/rpc/asio/encoder.hpp [hunk 1]
```diff
@@ -111,7 +112,23 @@ struct encoded:
         typedef typename event_traits<Event>::argument_type argument_type;
 
         type_traits<argument_type>::pack(packer, std::forward<Args>(args)...);
+        packer.pack_array(3);
+
+        auto trace_id = 0; //tracer::current_span()->get_trace_id();
```
Comment (3Hren) [inline include/cocaine/rpc/asio/encoder.hpp]: The type will be `int`, which is probably not what you want.

File: include/cocaine/rpc/asio/encoder.hpp [hunk 2]
```diff
@@ -111,7 +112,23 @@ struct encoded:
...
+        header_t::str trace_id_data{reinterpret_cast<const char*>(&trace_id), 8};
```
Comment (3Hren) [inline include/cocaine/rpc/asio/encoder.hpp]: This one looks weird, I don't understand what it does.

1. Is it necessary to use `reinterpret_cast`?
2. What that magic 8 means?

File: include/cocaine/rpc/asio/header.hpp [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+    Cocaine is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef HEADER_HPP
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: It's very likely, that we get naming conflict here. Please use full name with modules.

File: include/cocaine/rpc/asio/header.hpp [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+
+#include <cstring>
+#include <array>
+#include <cassert>
+#include <functional>
+
+namespace cocaine { namespace io  {
+
+enum class header_id {
+    TRACE_ID,
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Style, we use lowercase in enums.

File: include/cocaine/rpc/asio/header.hpp [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+    SPAN_ID,
+    PARENT_ID,
+    //Not used. Just indicates size of headers.
+    LAST
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Also not likely. Use some other way to indicate predefined headers size.

File: include/cocaine/rpc/asio/header.hpp [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+
+    inline
+    const std::string&
+    name() const {
+        return names()[static_cast<size_t>(id)];
+    }
+
+private:
+    header_id id;
+    static constexpr size_t header_cnt = static_cast<size_t>(header_id::LAST);
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Don't mangle variable names.

File: include/cocaine/rpc/asio/encoder.hpp (hunk 2 above)
Comment (antmat) [inline include/cocaine/rpc/asio/encoder.hpp]: headers are binary strings. here we pack uint64_t to binary string of 8 bytes. I've thought about adding a convinient method, but it seems it will be even less obvious that it packs numbers via binary representation.

File: include/cocaine/rpc/asio/header.hpp [hunk 7]
```diff
@@ -0,0 +1,558 @@
...
+
+    inline
+    static
+    const std::array<std::string, header_cnt>&
+    names() {
+        static std::array<std::string, header_cnt> header_names {{
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Here is the right place to get predefined headers size.

File: include/cocaine/rpc/asio/header.hpp [hunk 8]
```diff
@@ -0,0 +1,558 @@
...
+        return header_names;
+    }
+};
+
+// Header class.
+// Represents non-owning header referring to some external memory.
+// Can be constructed only via header table
+class header_t {
+public:
+    struct str {
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Is it **really** string, or maybe binary?

Also use `_t` suffix.

File: include/cocaine/rpc/asio/header.hpp [hunk 9]
```diff
@@ -0,0 +1,558 @@
...
+        const char* data;
+        size_t size;
+        bool operator==(const str& other) const {
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Blank line.

File: include/cocaine/rpc/asio/header.hpp [hunk 10]
```diff
@@ -0,0 +1,558 @@
...
+            return size == other.size && (memcmp(data, other.data, size) == 0);
+        }
+    };
+
+    header_t(const header_t& other) = default;
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: This one will result in warning on some ancient compilers when tried to emplace copy ctor somewhere.

File: include/cocaine/rpc/asio/header.hpp [hunk 11]
```diff
@@ -0,0 +1,558 @@
...
+    header_t& operator=(const header_t& other) = default;
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: And this one, JFYI.

File: include/cocaine/rpc/asio/header.hpp [hunk 12]
```diff
@@ -0,0 +1,558 @@
...
+        value(_value)
+    {}
+
+    inline
+    str
+    get_name() const;
+
+    inline
+    str
+    get_string_value() const;
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Value or representation? Also string or binary?

File: include/cocaine/rpc/asio/header.hpp [hunk 13]
```diff
@@ -0,0 +1,558 @@
...
+
+    // Just a convience conversion of chunk of data to uint64_t.
+    // Header has no type information, so
+    // it's up to user to determine what he expects to see - string or integer.
+    inline
+    uint64_t
+    get_numeric_value() const;
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Ditto.

File: include/cocaine/rpc/asio/header.hpp [hunk 14]
```diff
@@ -0,0 +1,558 @@
...
+
+    inline
+    header_t() = default;
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Place all constructors together.

File: include/cocaine/rpc/asio/header.hpp [hunk 15]
```diff
@@ -0,0 +1,558 @@
...
+    const header_t&
+    operator[](size_t idx);
+
+    inline
+    bool
+    parse_headers(std::vector<header_t>& parse_to, const msgpack::object& parse_from);
+
+    inline
+    header_t
+    push(const msgpack::object& data);
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: I suppose `header_table` should have no idea about out protocol implementation.

File: include/cocaine/rpc/asio/header.hpp [hunk 16]
```diff
@@ -0,0 +1,558 @@
...
+
+    inline
+    void
+    push(header_t& header);
+
+    template<class Packer>
+    void pack(Packer& packer, header_t header) {
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Request external packing.

File: include/cocaine/rpc/asio/header.hpp [hunk 17]
```diff
@@ -0,0 +1,558 @@
...
+
+    inline
+    size_t
+    size() const;
+
+    inline
+    bool
+    empty() const;
+
+    static constexpr size_t MAX_DATA_CAPACITY = 4096;
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Style, lowercase even for compile-time constants.

File: include/cocaine/rpc/asio/header.hpp [hunk 18]
```diff
@@ -0,0 +1,558 @@
...
+
+    //32 bytes overhead per record and 2 bytes for nil-nil header
+    static constexpr size_t MAX_HEADER_CAPACITY = MAX_DATA_CAPACITY/34;
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Ditto.

File: include/cocaine/rpc/asio/header.hpp [hunk 19]
```diff
@@ -0,0 +1,558 @@
...
+    static const std::vector<header_t>&
+    static_data();
+
+    inline
+    void
+    pop();
+
+    inline
+    size_t
+    find_full(const header_t& header) {
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: `full` what?

File: include/cocaine/rpc/asio/header.hpp [hunk 20]
```diff
@@ -0,0 +1,558 @@
...
+
+    // Header storage. Implemented as circular buffer
+    std::array<header_t, MAX_HEADER_CAPACITY> headers;
+    size_t header_lower_bound;
+    size_t header_upper_bound;
+
+    // Header data storage. Stores all data which headers can reference.
+    // Implemented as a sort of circular buffer.
+    // We multiply by 2 as data can be padded and we don't want to move it in memory
+    std::array<char, MAX_DATA_CAPACITY*2> header_data;
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Why `2`, not `3`?

File: include/cocaine/rpc/asio/header.hpp [hunk 21]
```diff
@@ -0,0 +1,558 @@
...
+    data_upper_bound(0),
+    capacity(MAX_DATA_CAPACITY)
+{}
+
+size_t
+header_table_t::size() const {
+    if(data_upper_bound > data_lower_bound) {
+        return static_data().size() + data_upper_bound - data_lower_bound;
+    }
+    else {
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Style.

File: include/cocaine/rpc/asio/header.hpp [hunk 22]
```diff
@@ -0,0 +1,558 @@
...
+}
+
+
+header_t
+header_table_t::push(const msgpack::object& data) {
+    header_t result;
+    //If header is fully from the table just fill it and return
+    if(data.type == msgpack::type::POSITIVE_INTEGER) {
+        if(data.via.u64 >= size() || data.via.u64 == 0) {
+            throw cocaine::error_t("Invalid index for header table");
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Style for the first char.

File: include/cocaine/rpc/asio/header.hpp [hunk 23]
```diff
@@ -0,0 +1,558 @@
...
+        }
+        result = operator [](data.via.u64);
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: `operator []` it's a name, no whitespaces required.

File: include/cocaine/rpc/asio/header.hpp [hunk 24]
```diff
@@ -0,0 +1,558 @@
...
+    return result;
+}
+
+void
+header_table_t::push(header_t& result) {
+
+    // Proceed to encode header to table
+    size_t value_size_size = http2_integer_size(result.value.size, 1);
+    size_t value_size = result.value.size;
+    size_t header_size = 32 + result.name.size + http2_integer_size(result.name.size, 1) + value_size_size + value_size;
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Magic 32?

File: include/cocaine/rpc/asio/header.hpp [hunk 25]
```diff
@@ -0,0 +1,558 @@
...
+
+    //pop headers from table until there is enough room for new one
+    while(data_upper_bound - data_lower_bound + header_size > capacity && data_lower_bound != data_upper_bound) {
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Too complex logic, consider moving it to the well-named method.

File: include/cocaine/rpc/asio/header.hpp [hunk 26]
```diff
@@ -0,0 +1,558 @@
...
+}
+
+#define COCAINE_MAKE_STATIC_HEADER(name, val) header_t(name, sizeof(name)-1, val, sizeof(val)-1)
+#define COCAINE_MAKE_STATIC_EMPTY_HEADER(name) COCAINE_MAKE_STATIC_HEADER(name, "")
+const std::vector<header_t>&
+header_table_t::static_data() {
+    static std::vector<header_t> headers ({
+        //Empty header is reserved as 0 is not valid index in http2
+        COCAINE_MAKE_STATIC_EMPTY_HEADER(""),
+        COCAINE_MAKE_STATIC_HEADER("trace_id", "\0\0\0\0\0\0\0\0"),
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: No macros please.

File: include/cocaine/rpc/asio/header.hpp [hunk 27]
```diff
@@ -0,0 +1,558 @@
...
+}
+
+bool
+operator==(const header_t::str& str, const header_key_t& header_key) {
+    return operator==(header_key, str);
+}
+
+size_t
+http2_integer_size(size_t sz, size_t bit_offset) {
+    if(sz < (1 << (8 - bit_offset))) {
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Some comments are required here.

File: include/cocaine/rpc/asio/header.hpp [hunk 28]
```diff
@@ -0,0 +1,558 @@
...
+        sz = sz << 7;
+        ret++;
+    }
+    return ret;
+}
+
+//Note dest should be at least 10 bytes long
+size_t
+http2_integer_encode(char* dest, size_t source, size_t bit_offset, char prefix) {
+    dest[0] = prefix;
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: And here(comments).

File: src/session.cpp [hunk 29]
```diff
@@ -129,14 +129,15 @@ session_t::push_action_t::operator()(const std::shared_ptr<channel<tcp>> ptr) {
 
 void
 session_t::push_action_t::finalize(const std::error_code& ec) {
-    if(ec.value() == 0) return;
```
Comment (3Hren) [inline src/session.cpp]: Do not change code style in already existing files, please.

File: src/session.cpp [hunk 30]
```diff
@@ -129,14 +129,15 @@ session_t::push_action_t::operator()(const std::shared_ptr<channel<tcp>> ptr) {
...
+    if(ec.value() == 0) {
+        return;
+    };
 
     if(ec != asio::error::eof) {
         COCAINE_LOG_ERROR(session->log, "client disconnected: [%d] %s", ec.value(), ec.message());
     } else {
         COCAINE_LOG_DEBUG(session->log, "client disconnected");
     }
-
```
Comment (3Hren) [inline src/session.cpp]: The same.

File: src/session.cpp [hunk 31]
```diff
@@ -167,7 +168,6 @@ session_t::session_t(std::unique_ptr<logging::log_t> log_,
 void
 session_t::handle(const decoder_t::message_type& message) {
     const channel_map_t::key_type channel_id = message.span();
-
```
Comment (3Hren) [inline src/session.cpp]: Ditto.

File: include/cocaine/rpc/asio/decoder.hpp [hunk 32]
```diff
@@ -49,8 +51,20 @@ struct decoded_message_t {
         return object.via.array.ptr[2];
     }
 
+    boost::optional<header_t>
+    get_header(const header_key_t& key) const {
```
Comment (3Hren) [inline include/cocaine/rpc/asio/decoder.hpp]: Where do we need this method?

File: include/cocaine/rpc/asio/encoder.hpp [hunk 33]
```diff
@@ -22,6 +22,7 @@
 #define COCAINE_IO_ENCODER_HPP
 
 #include "cocaine/rpc/protocol.hpp"
+#include "cocaine/rpc/asio/header.hpp"
```
Comment (3Hren) [inline include/cocaine/rpc/asio/encoder.hpp]: Include order.

File: include/cocaine/rpc/asio/decoder.hpp [hunk 34]
```diff
@@ -25,6 +25,8 @@
 
 #include "cocaine/traits.hpp"
 
+#include "cocaine/rpc/asio/header.hpp"
```
Comment (3Hren) [inline include/cocaine/rpc/asio/decoder.hpp]: Include order.

File: include/cocaine/rpc/asio/encoder.hpp [hunk 35]
```diff
@@ -111,7 +112,23 @@ struct encoded:
         typedef typename event_traits<Event>::argument_type argument_type;
 
         type_traits<argument_type>::pack(packer, std::forward<Args>(args)...);
+        packer.pack_array(3);
```
Comment (3Hren) [inline include/cocaine/rpc/asio/encoder.hpp]: Why don't you pack with `type_traits<T>::pack` here?

File: include/cocaine/rpc/asio/header.hpp [hunk 36]
```diff
@@ -0,0 +1,558 @@
...
+    along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef HEADER_HPP
+#define HEADER_HPP
+
+#include "cocaine/traits.hpp"
+#include "cocaine/exceptions.hpp"
+
+#include <cstring>
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Sort. Also drop excess includes if there are some.

File: include/cocaine/rpc/asio/header.hpp (hunk 9 above)
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Also the type should be in a separate line.

File: include/cocaine/rpc/asio/header.hpp (hunk 11 above)
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Also style.

File: include/cocaine/rpc/asio/header.hpp [hunk 37]
```diff
@@ -0,0 +1,558 @@
...
+
+    inline
+    str
+    get_string_value() const;
+
+    // Just a convience conversion of chunk of data to uint64_t.
+    // Header has no type information, so
+    // it's up to user to determine what he expects to see - string or integer.
+    inline
+    uint64_t
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Looks highly type unsafe. Consider external cast function for this purposes.

File: include/cocaine/rpc/asio/header.hpp [hunk 38]
```diff
@@ -0,0 +1,558 @@
...
+    get_numeric_value() const;
+
+    inline
+    header_t() = default;
+
+    bool operator==(const header_t& other) const {
+        return name == other.name && value == other.value;
+    }
+
+    bool name_equal(const header_t& other) const {
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Why do we need this method?

File: include/cocaine/rpc/asio/header.hpp [hunk 39]
```diff
@@ -0,0 +1,558 @@
...
+    // Header has no type information, so
+    // it's up to user to determine what he expects to see - string or integer.
+    inline
+    uint64_t
+    get_numeric_value() const;
+
+    inline
+    header_t() = default;
+
+    bool operator==(const header_t& other) const {
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Result type style.

File: include/cocaine/rpc/asio/header.hpp [hunk 40]
```diff
@@ -0,0 +1,558 @@
...
+                    //Either raw data or a number.
+                    obj.via.array.ptr[2].type == msgpack::type::RAW ||
+                    obj.via.array.ptr[2].type == msgpack::type::POSITIVE_INTEGER
+               )
+           )
+        ) {
+            try {
+                parse_to.push_back(push(obj));
+            }
+            catch(const cocaine::error_t& e) {
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Style.

File: include/cocaine/rpc/asio/header.hpp [hunk 41]
```diff
@@ -0,0 +1,558 @@
...
+        COCAINE_MAKE_STATIC_HEADER("span_id", "\0\0\0\0\0\0\0\0"),
+        COCAINE_MAKE_STATIC_HEADER("parent_id", "\0\0\0\0\0\0\0\0")
+    });
+
+    return headers;
+}
+
+bool
+operator==(const header_key_t& header_key, const header_t::str& str) {
+    //fast check first
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Operator &&?

File: include/cocaine/rpc/asio/header.hpp (hunk 10 above)
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: It's not **really** copyable, but its okay.

File: include/cocaine/rpc/asio/header.hpp (hunk 40 above)
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Also too broad exception to be eaten.

File: include/cocaine/rpc/asio/decoder.hpp (hunk 32 above)
Comment (antmat) [inline include/cocaine/rpc/asio/decoder.hpp]: we don't use headers now at all, however we need a method to fetch them from the message.

File: include/cocaine/api/isolate.hpp [hunk 42]
```diff
@@ -86,13 +86,14 @@ struct category_traits<isolate_t> {
         get(context_t& context, const std::string& name, const dynamic_t& args) {
             ptr_type instance;
 
-            auto lock_ptr = instances.synchronize();
```
Comment (3Hren) [inline include/cocaine/api/isolate.hpp]: This change is out of scope of this PR. Moreover I'm pretty sure I've already seen these changes in 99b34ca8786a181e1773ed1ddf65f5122a2a77a2

File: include/cocaine/api/storage.hpp [hunk 43]
```diff
@@ -139,13 +139,14 @@ struct category_traits<storage_t> {
         get(context_t& context, const std::string& name, const dynamic_t& args) {
             ptr_type instance;
 
-            auto lock_ptr = instances.synchronize();
```
Comment (3Hren) [inline include/cocaine/api/storage.hpp]: The same.

File: include/cocaine/rpc/asio/encoder.hpp (hunk 33 above)
Comment (3Hren) [inline include/cocaine/rpc/asio/encoder.hpp]: The issue still exists.

File: include/cocaine/rpc/asio/encoder.hpp (hunk 35 above)
Comment (3Hren) [inline include/cocaine/rpc/asio/encoder.hpp]: Ping.

File: include/cocaine/rpc/asio/header.hpp (hunk 36 above)
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Ping.

File: include/cocaine/rpc/asio/header.hpp [hunk 44]
```diff
@@ -0,0 +1,545 @@
...
+#include <boost/mpl/for_each.hpp>
+
+namespace cocaine { namespace io  {
+namespace header {
+
+struct data_t {
+    const char* blob;
+    size_t size;
+    bool operator==(const data_t& other) const {
+        return size == other.size && (memcmp(blob, other.blob, size) == 0);
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: `::memcmp` or `std::memcmp`

File: include/cocaine/rpc/asio/header.hpp [hunk 45]
```diff
@@ -0,0 +1,545 @@
...
+    }
+
+    bool
+    name_equal(const header_t& other) const {
+        return name == other.name;
+    }
+
+    friend struct header_traits;
+    friend struct header_static_table_t;
+private:
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Blank line before `private:`.

File: include/cocaine/rpc/asio/header.hpp [hunk 46]
```diff
@@ -0,0 +1,545 @@
...
+
+        static
+        constexpr
+        header::data_t
+        value() {
+            return DefaultValue::value();
+        }
+    };
+
+    template <class DefaultValue = default_values_t::zero_uint_value_t>
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Style.

File: include/cocaine/rpc/asio/header.hpp [hunk 47]
```diff
@@ -0,0 +1,545 @@
...
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Style.

File: include/cocaine/rpc/asio/header.hpp [hunk 48]
```diff
@@ -0,0 +1,545 @@
...
+    size_t
+    idx() {
+        static_assert(boost::mpl::contains<headers, Header>::type::value, "Could not find header in statis table");
+        return boost::mpl::find<headers, Header>::type::pos::value;
+    }
+
+private:
+    template<size_t Size>
+    struct init_header {
+    template<class Header>
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Indent.

File: include/cocaine/rpc/asio/header.hpp [hunk 49]
```diff
@@ -0,0 +1,545 @@
...
+    constexpr
+    static
+    size_t
+    idx() {
+        static_assert(boost::mpl::contains<headers, Header>::type::value, "Could not find header in statis table");
+        return boost::mpl::find<headers, Header>::type::pos::value;
+    }
+
+private:
+    template<size_t Size>
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Why this template parameter is required?

File: include/cocaine/rpc/asio/header.hpp [hunk 50]
```diff
@@ -0,0 +1,545 @@
...
+    template<class Header>
+        void
+        operator()(Header) {
+            data[boost::mpl::find<headers, Header>::type::pos::value].name = Header::name();
+            data[boost::mpl::find<headers, Header>::type::pos::value].value = Header::value();
+        }
+        std::array<header_t, size> data;
+    };
+
+    struct placeholder_t {
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Where this one is used.

File: include/cocaine/rpc/asio/header.hpp [hunk 51]
```diff
@@ -0,0 +1,545 @@
...
+        // Used to suppress unused variable notice
+        placeholder_t() {}
+        placeholder_t(const placeholder_t&) {}
+        placeholder_t& operator=(const placeholder_t&) {return *this;}
+    };
+
+    static
+    std::array<header_t, size>
+    init_data() {
+        init_header<size> initer;
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: `initer`?

File: include/cocaine/rpc/asio/header.hpp [hunk 52]
```diff
@@ -0,0 +1,545 @@
...
+    size() const;
+
+    inline
+    bool
+    empty() const;
+
+    static constexpr size_t max_data_capacity = 4096;
+    static constexpr size_t http2_header_overhead = 32;
+    //32 bytes overhead per record and 2 bytes for nil-nil header
+    static constexpr size_t max_header_capacity = max_data_capacity/(http2_header_overhead + 2);
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Spaces between `/`.

File: include/cocaine/rpc/asio/header.hpp [hunk 53]
```diff
@@ -0,0 +1,545 @@
...
+    data_lower_bound(0),
+    data_upper_bound(0),
+    capacity(max_data_capacity)
+{}
+
+size_t
+header_table_t::size() const {
+    if(data_upper_bound > data_lower_bound) {
+        return header_static_table_t::size + data_upper_bound - data_lower_bound;
+    }
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Style.

File: include/cocaine/rpc/asio/header.hpp [hunk 54]
```diff
@@ -0,0 +1,545 @@
...
+}
+
+bool
+header_table_t::empty() const {
+    return data_lower_bound == data_upper_bound;
+}
+
+void
+header_table_t::push(header_t& result) {
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: Document, decompose or do something with this method. I don't understand how it works.

File: include/cocaine/traits/header.hpp [hunk 55]
```diff
@@ -0,0 +1,159 @@
...
+    // Pack a header from static table with predefined value
+    template<class Header, class Stream>
+    static
+    void
+    pack(msgpack::packer<Stream>& packer) {
+        static_assert(boost::mpl::contains<header_static_table_t::headers, Header>::type::value, "Header is not present in static table");
+        packer.pack_fix_uint64(header_static_table_t::idx<Header>());
+    }
+
+    // Pack a header from statis table but with different value
```
Comment (3Hren) [inline include/cocaine/traits/header.hpp]: stati**s**

File: include/cocaine/traits/header.hpp [hunk 56]
```diff
@@ -0,0 +1,159 @@
...
+    template<class Header, class Stream>
+    static
+    void
+    pack(msgpack::packer<Stream>& packer, header_table_t& table, header::data_t& header_data) {
```
Comment (3Hren) [inline include/cocaine/traits/header.hpp]: Are you sure you need non-const reference for packing?

File: include/cocaine/traits/header.hpp [hunk 57]
```diff
@@ -0,0 +1,159 @@
...
+        packer.pack_fix_uint64(pos);
+        packer.pack_raw(header_data.size);
+        packer.pack_raw_body(header_data.blob, header_data.size);
+    }
+
+    // Pack any other header
+    template<class Stream>
+    static
+    void
+    pack(msgpack::packer<Stream>& packer, header_table_t& table, header_t& source) {
```
Comment (3Hren) [inline include/cocaine/traits/header.hpp]: Ditto.

File: src/session.cpp [hunk 58]
```diff
@@ -302,7 +302,8 @@ session_t::push(encoder_t::message_type&& message) {
     if(const auto ptr = *transport.synchronize()) {
 #endif
         // Use dispatch() instead of a direct call for thread safety.
-        ptr->socket->get_io_service().dispatch(std::bind(&push_action_t::operator(),
+        ptr->socket->get_io_service().dispatch(std::bind(
```
Comment (3Hren) [inline src/session.cpp]: Out of the scope.

**Comment by 3Hren** [comment]
Almost everything is ok except some minor style issues.

But it's nontrivial to understand how the header table works, so I'd like to see several tests to be sure that everything is ok.

File: include/cocaine/traits/header.hpp (hunk 56 above)
Comment (antmat) [inline include/cocaine/traits/header.hpp]: Yes, because table is modified and header data starts to point to header data.

**Comment by 3Hren** [comment]
@noxiouz, you can also take a look, if you want.

File: include/cocaine/rpc/asio/header.hpp [hunk 59]
```diff
@@ -0,0 +1,631 @@
...
+        init_header_t() :
+            data(std::make_shared<storage_t>())
+        {}
+        template<class Header>
+        void
+        operator()(Header) {
+            (*data)[boost::mpl::find<headers, Header>::type::pos::value].name = Header::name();
+            (*data)[boost::mpl::find<headers, Header>::type::pos::value].value = Header::value();
+        }
+        //this is because mpl::for_each copies functor;
```
Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]: O_o

Comment (3Hren) [inline include/cocaine/rpc/asio/header.hpp]:
```
template<class Header>
void operator()(Header*) { ... }

boost::mpl::for_each<boost::mpl::transform<headers, std::add_pointer<boost::mpl::_1>>::type>(init);
```

**Comment by noxiouz** [comment]
@3Hren Thanks! I'm gonna do it in one hour.

File: include/cocaine/rpc/asio/header.hpp [hunk 60]
```diff
@@ -0,0 +1,631 @@
...
+header_t::http2_size() const {
+    // 1 refer to string literals which has size with 1-bit padding.
+    // See https://tools.ietf.org/html/draft-ietf-httpbis-header-compression-12#section-5.2
+    return name.size + http2_integer_size(name.size, 1) + value.size + http2_integer_size(value.size, 1) + header_table_t::http2_header_overhead;
+}
+
+header_table_t::header_table_t() :
+    header_lower_bound(0),
+    data_lower_bound_end(0),
+    header_upper_bound(0),
```
Comment (noxiouz) [inline include/cocaine/rpc/asio/header.hpp]: There'are defined in the class in another order
```cpp
header_lower_bound(0),
data_lower_bound_end(0), 
header_upper_bound(0),
```

File: include/cocaine/rpc/asio/header.hpp [hunk 61]
```diff
@@ -0,0 +1,631 @@
...
+
+    // Header do not fit in the table. According to RFC we just clean the table and do not put the header inside.
+    if(empty() && data_size() + header_size > capacity) {
+        return;
+    }
+
+    // Find the appropriate position in header table.
+    char* dest = header_data.data();
+    if(header_data.size() - data_upper_bound < header_size) {
+        // If header do not fit in the remaining space in the end of buffer - we write at the beginning.
```
Comment (noxiouz) [inline include/cocaine/rpc/asio/header.hpp]: does not

**Comment by 3Hren** [comment]
LGFM

**Comment by 3Hren** [comment]
Reverted due to:
 - Framework API/ABI breakage without providing the replacement.
 - Phantom headers.
 - Improper headers orders.
 - Useless traffic (see any Cocaine response).
 - New codestyle issues.
